sourced/ml/transformers/basic.py

```diff
@@ -163,33 +166,45 @@ def __call__(self, head: RDD):
 
 
 class LanguageSelector(Transformer):
-    def __init__(self, languages: Union[list, tuple], blacklist=False, **kwargs):
+    def __init__(self, languages: list, blacklist=False, **kwargs):
         super().__init__(**kwargs)
         self.languages = languages
         self.blacklist = blacklist
 
     def __call__(self, files: DataFrame) -> DataFrame:
-        files = files.dropDuplicates(("blob_id",)).filter("is_binary = 'false'")
-        classified = files.classify_languages()
-        if not self.blacklist:
-            lang_filter = classified.lang == self.languages[0]
-            for lang in self.languages[1:]:
-                lang_filter |= classified.lang == lang
-        else:
-            lang_filter = classified.lang != self.languages[0]
-            for lang in self.languages[1:]:
-                lang_filter &= classified.lang != lang
-        filtered_by_lang = classified.filter(lang_filter)
-        return filtered_by_lang
+        if not isinstance(files, BlobsDataFrame):
+            self._log.warning("Argument type is not BlobsDataFrame. "
+                              "Language classification will be skipped.")
+        elif "lang" not in files.columns:
+            files = files \
+                .dropDuplicates(("blob_id",)) \
+                .filter("is_binary = 'false'") \
+                .classify_languages()
+        return files[files.lang.isin(self.languages) != self.blacklist]
+
+    @staticmethod
+    def maybe(languages, blacklist):
+        if languages is None:
+            return Identity()
+        return LanguageSelector(languages, blacklist)
 
 
 class UastExtractor(Transformer):
     def __init__(self, **kwargs):
         super().__init__(**kwargs)
 
     def __call__(self, files: DataFrame) -> DataFrame:
-        # if UAST is not extracted, returns an empty list that we filter out here
-        return files.extract_uasts().where(functions.size(functions.col("uast")) > 0)
+        if isinstance(files, (BlobsDataFrame, BlobsWithLanguageDataFrame)):
```

Comment:
not really, the problem here is in parquet files. If you start with them, UAST can already be extracted. Of course, we can check it in the middle of the pipeline, but I prefer to move this check to this place because of our functions like `create_uast_source` were extremely complex, branched and confusing. Now it is only 3 lines.

Reply:
Refactor, nuke, rewrite but I need UastExtractor to extract uasts and accept the single data type RDD. The rest goes to the upper scope.